File: packages/typescript/src/envelope.ts
import type { YaiStatus } from './status';

export interface YaiEnvelope<TData = unknown> {
  operation_id: string;
  status: YaiStatus;
  execution_claim: boolean;
  implementation_status: string;
  system_call_ref?: string;
  client_subject_ref?: string;
  client_connection_ref?: string;
  client_attachment_ref?: string;
  system_root_context_ref?: string;
  work_case_ref?: string;
  control_admission_ref?: string;
  message?: string;
  data?: TData;
}

File: packages/rust/src/envelope.rs
use serde::{Deserialize, Serialize};

use crate::status::YaiStatus;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct YaiEnvelope<TData> {
    pub operation_id: String,
    pub status: String,
    pub execution_claim: bool,
    pub implementation_status: String,
    pub system_call_ref: Option<String>,
    pub client_subject_ref: Option<String>,
    pub client_connection_ref: Option<String>,
    pub client_attachment_ref: Option<String>,
    pub system_root_context_ref: Option<String>,
    pub work_case_ref: Option<String>,
    pub control_admission_ref: Option<String>,
    pub message: Option<String>,
    pub data: Option<TData>,
}

impl<TData> YaiEnvelope<TData> {
    pub fn unavailable(operation_id: &'static str, message: impl Into<String>) -> Self {
        Self {
            operation_id: operation_id.to_string(),
            status: YaiStatus::Unavailable.as_str().to_string(),
            execution_claim: false,
            implementation_status: "transport-unconfigured".to_string(),
            system_call_ref: None,
            client_subject_ref: None,
            client_connection_ref: None,
            client_attachment_ref: None,
            system_root_context_ref: None,
            work_case_ref: None,
            control_admission_ref: None,
            message: Some(message.into()),
            data: None,
        }
    }
}
